### ISSUE TITLE
TTT bug in hypothesis model.

### ISSUE THREAD

**Issue by fhowar** [LearnLib/learnlib#77]
*My Environment*

- Learnlib Version 0.14.0 and 0.16.0 at least.
- Model: ```BitVise.dot``` [https://automata.cs.ru.nl/BenchmarkSSH/Mealy](https://automata.cs.ru.nl/BenchmarkSSH/Mealy)

*Problem*

In some experiments I make a copy of the hypothesis for finding counterexamples:

```java
CompactMealy<I,O> h = new CompactMealy<>(alphabet);
AutomatonLowLevelCopy.copy(AutomatonCopyMethod.DFS, hypothesis, alphabet, h);
```

This seems to work as expected for all learning algorithms except for TTT, where it sometimes results in 

```java
java.lang.NullPointerException
	at net.automatalib.automata.helpers.SimpleStateIDs.getStateId(SimpleStateIDs.java:44)
	at net.automatalib.automata.helpers.StateIDStaticMapping.get(StateIDStaticMapping.java:34)
	at net.automatalib.util.automata.copy.AbstractLowLevelAutomatonCopier.copyTransitionChecked(AbstractLowLevelAutomatonCopier.java:105)
	at net.automatalib.util.automata.copy.TraversalAutomatonCopy.processTransition(TraversalAutomatonCopy.java:81)
	at net.automatalib.util.ts.traversal.TSTraversal.depthFirst(TSTraversal.java:102)
	at net.automatalib.util.ts.traversal.TSTraversal.traverse(TSTraversal.java:244)
	at net.automatalib.util.automata.copy.TraversalAutomatonCopy.doCopy(TraversalAutomatonCopy.java:56)
	at net.automatalib.util.automata.copy.AutomatonLowLevelCopy.rawCopy(AutomatonLowLevelCopy.java:139)
	at net.automatalib.util.automata.copy.AutomatonLowLevelCopy.copy(AutomatonLowLevelCopy.java:340)
	at net.automatalib.util.automata.copy.AutomatonLowLevelCopy.copy(AutomatonLowLevelCopy.java:525)
	at net.automatalib.util.automata.copy.AutomatonLowLevelCopy.copy(AutomatonLowLevelCopy.java:602)
	at net.automatalib.util.automata.copy.AutomatonLowLevelCopy.copy(AutomatonLowLevelCopy.java:561)
	at de.learnlib.cav2020.performance.OptimalExamples.generateCounterexample(OptimalExamples.java:170)
	at de.learnlib.cav2020.performance.OptimalExamples.runLearner(OptimalExamples.java:142)
	at de.learnlib.cav2020.performance.OptimalExamples.runExample(OptimalExamples.java:315)
	at de.learnlib.cav2020.performance.OptimalExamples.runMain(OptimalExamples.java:493)
	at com.misberner.clitools.CLIToolDispatcher.runTool(CLIToolDispatcher.java:74)
	at com.misberner.clitools.CLIToolDispatcher.run(CLIToolDispatcher.java:63)
	at de.learnlib.cav2020.LearnlibTool.main(LearnlibTool.java:26)
```

or

```java
java.lang.NullPointerException
	at de.learnlib.algorithms.ttt.base.AbstractTTTLearner.splitState(AbstractTTTLearner.java:241)
	at de.learnlib.algorithms.ttt.base.AbstractTTTLearner.splitState(AbstractTTTLearner.java:279)
	at de.learnlib.algorithms.ttt.base.AbstractTTTLearner.refineHypothesisSingle(AbstractTTTLearner.java:212)
	at de.learnlib.algorithms.ttt.mealy.TTTLearnerMealy.refineHypothesisSingle(TTTLearnerMealy.java:67)
	at de.learnlib.algorithms.ttt.base.AbstractTTTLearner.refineHypothesis(AbstractTTTLearner.java:166)
	at de.learnlib.cav2020.performance.OptimalExamples.runLearner(OptimalExamples.java:148)
	at de.learnlib.cav2020.performance.OptimalExamples.runExample(OptimalExamples.java:315)
	at de.learnlib.cav2020.performance.OptimalExamples.runMain(OptimalExamples.java:493)
	at com.misberner.clitools.CLIToolDispatcher.runTool(CLIToolDispatcher.java:74)
	at com.misberner.clitools.CLIToolDispatcher.run(CLIToolDispatcher.java:63)
	at de.learnlib.cav2020.LearnlibTool.main(LearnlibTool.java:26)
```

*More Details*

I use the following code to generate long counterexamples. The problem occurs from upwards of length 40 at least

```java

	@SuppressWarnings("unchecked")
	private static void setInitialState(CompactMealy a, Object q0) {
		a.setInitialState(q0);
	}

	private static <I, O> Word<I>
	generateCounterexample(Random random, MealyLearningExample<I,O> example, MealyMachine<?,I,?,O> hypothesis, int ceLength) {

		MealyMachine<?,I,?,O> target = example.getReferenceAutomaton();
		Alphabet<I> alphabet = example.getAlphabet();

		CompactMealy<I,O> h = new CompactMealy<>(alphabet);
		AutomatonLowLevelCopy.copy(AutomatonCopyMethod.DFS, hypothesis, alphabet, h);
		CompactMealy<I,O> t = new CompactMealy<>(alphabet);
		AutomatonLowLevelCopy.copy(AutomatonCopyMethod.DFS, target, alphabet, t);


		Object hyp0 = h.getInitialState();
		Object sys0 = t.getInitialState();

		Word<I> ce = Automata.findSeparatingWord(t, h, alphabet);
		if(ce == null) {
			return ce;
		}

		Word<I> word = null;
		for (int i=0; i<2; i++) {
			WordBuilder<I> wb = new WordBuilder<>();			
			for(int j = 0; j < ceLength; j++) {
				I input = alphabet.getSymbol(random.nextInt(alphabet.size()));
				Word test = wb.toWord();
				test = test.append(input);

				Object hypTemp = h.getState(test);
				Object sysTemp = t.getState(test);
				setInitialState( h, hypTemp);
				setInitialState( t, sysTemp);			
				Word middle = Automata.findSeparatingWord(t, h, alphabet);
				setInitialState( h, hyp0);
				setInitialState( t, sys0);
				if (middle != null) {
					wb.append(input);
				}
			}
			Word prefix = wb.toWord();
			Object hypTemp = h.getState(prefix);
			Object sysTemp = t.getState(prefix);

			setInitialState( h, hypTemp);
			setInitialState( t, sysTemp);			
			Word middle = Automata.findSeparatingWord(t, h, alphabet);
			setInitialState( h, hyp0);
			setInitialState( t, sys0);

			wb.append(middle.prefix(middle.length()));

			word = wb.toWord();
			Word<O> out1 = target.computeOutput(word); //.suffix(ceLength - minLength);
			Word<O> out2 = hypothesis.computeOutput(word); //.suffix(ceLength - minLength);
			if ( !out1.equals(out2) ) {
				int k = 0;
				while (out1.getSymbol(k).equals(out2.getSymbol(k))) {
					k++;
				}
				word = word.prefix(k+1);
				break;
			}
			word = null;
		}

		if (word == null) {
			word = ce;
		}
		return word;
	}
```

**Comment by fhowar** [comment]
Some more details for reproducing this. Sequence of counterexamples.

- CE1: CH_OPEN CH_CLOSE
- CE2: CH_DATA CH_DATA KEXINIT_PROCEED SERVICE_REQUEST_CONN CH_OPEN
- CE3: CH_EOF CH_DATA CH_REQUEST_PTY KEXINIT CH_REQUEST_PTY KEX30
- CE4: CH_EOF CH_CLOSE CH_DATA CH_DATA CH_CLOSE KEXINIT_PROCEED CH_DATA CH_REQUEST_PTY KEX30 NEWKEYS SERVICE_REQUEST_AUTH SERVICE_REQUEST_CONN SERVICE_REQUEST_AUTH SERVICE_REQUEST_AUTH KEXINIT_PROCEED SERVICE_REQUEST_CONN KEXINIT_PROCEED KEX30 SERVICE_REQUEST_CONN CH_EOF CH_OPEN CH_OPEN
- CE5: CH_REQUEST_PTY CH_CLOSE KEXINIT CH_REQUEST_PTY KEX30 KEX30 NEWKEYS CH_OPEN CH_REQUEST_PTY CH_OPEN CH_EOF CH_CLOSE CH_REQUEST_PTY SERVICE_REQUEST_CONN

**Comment by fhowar** [comment]
My bad. I had not configured dependencies properly. Seems to be fixed in 0.16.0
